sdk/identity/Azure.Identity/src/DefaultAzureCredential.cs

```diff
@@ -90,53 +92,80 @@ public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext r
             return await GetTokenAsync(true, requestContext, cancellationToken).ConfigureAwait(false);
         }
 
-        private async Task<AccessToken> GetTokenAsync(bool isAsync, TokenRequestContext requestContext, CancellationToken cancellationToken)
+        private async ValueTask<AccessToken> GetTokenAsync(bool isAsync, TokenRequestContext requestContext, CancellationToken cancellationToken)
         {
             using CredentialDiagnosticScope scope = _pipeline.StartGetTokenScope("DefaultAzureCredential.GetToken", requestContext);
 
-            List<Exception> exceptions = new List<Exception>();
+            try
+            {
+                AccessToken token;
+
+                if (_credential != null)
+                {
+                    token = isAsync ? await _credential.GetTokenAsync(requestContext, cancellationToken).ConfigureAwait(false) : _credential.GetToken(requestContext, cancellationToken);
+                }
+                else
+                {
+                    token = isAsync ? await GetTokenFromSourcesAsync(isAsync, requestContext, cancellationToken).ConfigureAwait(false) : GetTokenFromSourcesAsync(isAsync, requestContext, cancellationToken).GetAwaiter().GetResult();
+                }
+
+                return scope.Succeeded(token);
+            }
+            catch (OperationCanceledException e)
+            {
+                scope.Failed(e);
+
+                throw;
+            }
+            catch (Exception e) when (!(e is CredentialUnavailableException))
+            {
+                throw scope.Failed(new AuthenticationFailedException(UnhandledExceptionMessage, e));
```

Comment:
Can we throw from `scope.Failed` implementation?

Reply:
Not sure if it matters.

Reply:
There is nuance with how these exceptions are thrown, for instance `OperationCanceledException` is thrown without being wrapped. I think this nuance would be hard to preserve if I moved the throw into `scope.Failed`

Reply:
Based off our offline discussion I'm going to break `scope.Failed` into two methods. One which only fails the `DiagnosticScope` will still be named `Failed`, and the one which wraps the exception in an `AuthenticationFailedException` will be called `FailAndWrap` this way it is more evident what the behavior is at the callsite.